### PR TITLE
Distribute shadow map update over multiple frames to reduce stutter

### PR DESCRIPTION
**Pull request by x2048**
Reduces stutter and rendering freezes when playing. Mitigation of #11357.

This PR distributes generation of ClientMap SM texture over multiple frames.
1. Maintains double SM and SM Color textures
2. Light frustum update triggers incremental generation of shadow map into secondary 'future' textures.
3. Every incremental update renders a portion of the shadow draw list (split equally).
4. After defined number of frames (currently, 4), 'future' and 'current' textures are swapped, and DirectionalLight 'commits' the new frustum to use when rendering shadows on screen.

## To do

This PR is Work in Progress

- [x] Need help to resolve occasional flicker. Seems to be related to light frustum update.
- [x] Implement setting to configure the number of frames (from 1 to 16)
- [x] Connect light frustum updates to frames and remove shadow map update interval
- [x] Avoid duplicate textures when updating shadow map in one frame

## How to test

Set medium shadow quality, view range = 250, shadow range = 250, enable colored shadows.
Create a world and fly around.
Enable profiler and compare shadow metrics to primary rendering. Shadow draw list generation and shadow map rendering times must be smaller than primary draw list generation and rendering times.

### CODE REVIEW

**Comment by x2048** [comment]
@lhofhansl Can you help identifying the problem with flickering shadows?
@0xLiso This looks like a good solution to make PSM perform, can you take a look?

**Comment by hecktest** [comment]
-2ms frame time difference and jitter as low as 40% in a static shot - this speedhack actually works now.

**Comment by 0xLiso** [comment]
> @0xLiso This looks like a good solution to make PSM perform, can you take a look?

Yes @x2048  it reduces a lot the stutter. It's works fine :). About the frustum you are right there is some problems in it. I'm working on it. 
Great job :)

**Comment by lhofhansl** [comment]
I'll try later today.
So this obsoletes #11393? Is so let's please close that one.

**Comment by lhofhansl** [comment]
I see a marked improvement. The flickering is probably due to the camera offset change again.

File: src/client/shadows/dynamicshadowsrender.h
Comment (sfan5) [inline src/client/shadows/dynamicshadowsrender.h:120]: constants should be uppercase
also `constexpr`

File: src/client/clientmap.cpp
Comment (sfan5) [inline src/client/clientmap.cpp:660]: should `break` if the high bound is reached

File: src/client/shadows/dynamicshadowsrender.cpp
Comment (sfan5) [inline src/client/shadows/dynamicshadowsrender.cpp:369]: There is no point is passing these here, the variables are already available to the method

File: src/client/shadows/dynamicshadowsrender.cpp
Comment (sfan5) [inline src/client/shadows/dynamicshadowsrender.cpp:275]: `std::swap`

Comment (kilbith) [reply src/client/shadows/dynamicshadowsrender.cpp]: It can even be `std::move` as `tmp` is not reutilized afterward.

**Comment by x2048** [comment]
> I'll try later today.
> So this obsoletes #11393? Is so let's please close that one.

No, #11393 is completely independent and will shave 5ms off the max dtime, improving further.

**Comment by x2048** [comment]
@sfan5 Thanks for the review, I pushed fixes according to the comments.

**Comment by lhofhansl** [comment]
With this together with #11393 I now see some weird artifacts where the shadows change when the view angle changes.

**Comment by x2048** [comment]
@lhofhansl I've change the way shadow map list is split between frames. This resolved visual artifacts for me. Can you check?

**Comment by lhofhansl** [comment]
The artifacts are gone, but it seems to ignore shadow map update interval. :(

**Comment by x2048** [comment]
This is partially true. Because shadow map update is spread over 8 frames now, the actual interval will be a maximum of shadow update interval and time it takes to render 8 frames. A a result, in order to update shadows 5 times a second (update interval 0.2) you need to run the game at 40 FPS. Lower than that, shadows will begin to lag, essentially prioritizing main rendering. To me the effect was noticeable only when flying fast, making quick camera turns, or at time speed over 144.

**Comment by lhofhansl** [comment]
I see. That makes sense @x2048 , perhaps we want to remove map update interval option? And maybe replace it with the number of sections to use with values from 1-16 (default 8)...? On a slow machine you set it to 16, on a fast one you can set it to 1 or 2.

**Comment by lhofhansl** [comment]
There are still artifacts:
output

See that non-shadow that moves below the tree? That should not be there and it moves in the wrong direction (it's before noon)

**Comment by hecktest** [comment]
I agree that the update time setting should be changed to slice count, and that we still need an instant mode for whoever can afford it. (because this is still an ugly but effective speedhack)

In the latter case, please also disable the double buffering - no need to waste all that VRAM if it isn't being used.

**Comment by lhofhansl** [comment]
I found that when setting the map update interval to 1/20s (i.e. 0.05), the shadows move visually smooth. With my 60 fps that would translate to 3 slices. So an option might be to allow values from 2 to 16.
But I agree with @hecktest that for fast machines you'd want the shadow to be updated on each frame.

I am still concerns about the new artifacts, though.

**Comment by x2048** [comment]
I marked it as work in progress until I sort the artifacts out.

**Comment by lhofhansl** [comment]
Tangentially related... Is this code necessary: https://github.com/minetest/minetest/blob/master/src/client/shadows/dynamicshadows.cpp#L81
?

This transforms the frustum, clamps it to texel increments (per comment) and then transforms it back. I found that this part is responsible for a lot of the jitter. Simply removing this reduces the jitter significantly.

So the question is: Why do we need to clamp the frustum increments?

**Comment by 0xLiso** [comment]
> Tangentially related... Is this code necessary: https://github.com/minetest/minetest/blob/master/src/client/shadows/dynamicshadows.cpp#L81
> ?
> 
> This transforms the frustum, clamps it to texel increments (per comment) and then transforms it back. I found that this part is responsible for a lot of the jitter. Simply removing this reduces the jitter significantly.
> 


You are right, just removing the transformVect part the jitter almost disappear. Great catch!.

It's solved in my last PSM branch, plus some more features. But I'm waiting until this PR gets merge to send the other one.

> So the question is: Why do we need to clamp the frustum increments?

this is to clap the frustum movement to the Shadow Map texel density to avoid shadow flickering when camera moves.

**Comment by x2048** [comment]
@lhofhansl @hecktest I've reimplemented the way SM texture is generated using blend functions. The artifacts should be gone now. Can you take a look?

**Comment by lhofhansl** [comment]
@x2048 Any chance for rebasing after #11385 ?

**Comment by x2048** [comment]
@lhofhansl Yes, once I get some time :)

If it helps, I've merged all my PRs into https://github.com/x2048/minetest/tree/shadows-all

**Comment by lhofhansl** [comment]
Thanks for rebasing... @x2048 . Looks good now; I do not see any artifacts.

**Comment by lhofhansl** [comment]
Code looks good too, and the profiler shows that the rendercalls due to shadows are about 1/4 of the scene render calls. (transparent calls are the same, since they're done in slice 0).

So, as soon as the remaining tasks are done, this should be good to go, IMHO.

**Comment by x2048** [comment]
Great, thank you, @lhofhansl. I am planning to include [0xLiso:PSM](https://github.com/0xLiso/minetest-shadowmap/tree/PSM) tonight, which includes performance and quality fixes around shadow maps, see commit c017d33.

**Comment by lhofhansl** [comment]
@x2048 Cool. I think it might be better to do this and the PSM branch as two separate PRs.

File: client/shaders/nodes_shader/opengl_fragment.glsl
Comment (lhofhansl) [inline client/shaders/nodes_shader/opengl_fragment.glsl:177]: Wouldn't it be better to clamp this to [0, PCFBOUND]? As is it always does a little bit of filtering (0.5).

Comment (x2048) [reply client/shaders/nodes_shader/opengl_fragment.glsl]: This clamping seriously damages distant shadows, see commit and comment below.

**Comment by x2048** [comment]
@lhofhansl OK, I've now undone the merge and implemented the rest of the tasks, this PR is now ready for review.

I also found out that clamping the number of filter samples seriously damages remote shadows, especially on walls and cliffs. I've now increased the minimum number of filter samples to be 1/4 of the initial number, which gives visible performance improvement at the cost of small loss in quality of distant shadows

**Comment by lhofhansl** [comment]
Awesome. I'll have a look as soon as I'm back at a computer.

**Comment by x2048** [comment]
@lhofhansl @sfan5 @kilbith Going back to [0xLiso:PSM](https://github.com/0xLiso/minetest-shadowmap/tree/PSM) and c017d33, @0xLiso has been working on performance and quality of shadow rendering from another angle and tried to keep up with the changes in this PR, which takes effort, because there are overlapping changes in the same files.

From the other side, @lhofhansl expressed preference to keep individual patches in separate PRs, as a general practice, and I fully respect the position of the core developers, as you are the primary reviewers.

I do not think there's a need for a lengthy discussion here, so I'd propose the core developers to vote :+1: or :-1: to show support for merging the two patches.

From my side, I would prefer merging the patches to reduce our effort but will support any outcome.

As a side note, I am away for 4 days with only sporadic access to the Internet and no code, so I will pick it up late next week.

**Comment by lhofhansl** [comment]
Re: Merging. When I played with the PSM branch I saw a different set of artifacts. It's very hard to keep these apart. Perhaps we can get this one in quickly - it's ready as far as I can (I just have a question or two). And then get the changes from the PSM branch as general improvements. Just MHO. @sfan5 et al, please chime in.

**Comment by lhofhansl** [comment]
Tested. Looks good, and performance is better.

I noticed (and I think this is unrelated to this issues) that the bias seems to be wrong from zoom:
screenshot_20210719_124457

Notice how the player-shadow is non-contiguous near the top. No breaker, we adjust these things later.

File: src/client/clientmap.cpp
Comment (lhofhansl) [inline src/client/clientmap.cpp:658]: So we doing the transparent pass in the last frame? Why not the first frame? I.e. `if (is_transparent_pass && frame > 0) ...`

File: src/client/clientmap.cpp
Comment (lhofhansl) [inline src/client/clientmap.cpp:744]: Why do we need draw a line here?

File: src/client/shadows/dynamicshadowsrender.cpp
Comment (lhofhansl) [inline src/client/shadows/dynamicshadowsrender.cpp:515]: Interesting. I'll trust you on this one.

File: src/client/clientmap.cpp
Comment (x2048) [reply src/client/clientmap.cpp:658]: This is because there is no double buffering for colored shadows, but they still render with 'future' frustum. The very last frame is exactly before 'future' becomes 'current' and I avoid visual glitches. I think I'll need to document this and other subtle things carefully.

File: src/client/clientmap.cpp
Comment (x2048) [reply src/client/clientmap.cpp:744]: This is to force irrlicht to actually update internal state when I set the material, otherwise there is no effect of the preceding lines.

**Comment by x2048** [comment]
Yes, bias issues are not really related since no changes were made to that part. @0xLiso  has made some corrections in the PSM branch, so it probably belongs there.

File: src/client/clientmap.cpp
Comment (sfan5) [reply src/client/clientmap.cpp:744]: Add a comment for this, otherwise nobody will be able to remember.

File: builtin/settingtypes.txt
Comment (sfan5) [inline builtin/settingtypes.txt:621]:
```suggestion
#    Spread update of shadow map over this amount of frames.
```

**Review by sfan5 — APPROVED** [review]
Works fine.

File: src/client/shadows/dynamicshadowsrender.cpp
Comment (sfan5) [inline src/client/shadows/dynamicshadowsrender.cpp:266]:
```suggestion
			for (DirectionalLight &light : m_light_list)
```

File: src/client/shadows/dynamicshadowsrender.h
Comment (sfan5) [inline src/client/shadows/dynamicshadowsrender.h:129]: no point to use a bigger type since `m_current_frame <= m_map_shadow_update_frames`
```suggestion
	u8 m_current_frame{0}; /* Current frame */
```

**Review by SmallJoker — APPROVED** [review]

File: src/client/shadows/dynamicshadowsrender.h
Comment (x2048) [reply src/client/shadows/dynamicshadowsrender.h]: I'd suggest to put a condition when incrementing m_current_frame, to avoid surprises with overflow when conditions for updating light frustum change.

**Pull request merged**
